gtsam/sfm/PowerMethod.h

```diff
@@ -0,0 +1,367 @@
+
+ // Matrix-vector multiplication operation
+ Vector perform_op(const Vector& x1, const Vector& x0) const {
+   // Do the multiplication
+   Vector x2 = A_ * x1 - beta_ * x0;
+   x2.normalize();
+   return x2;
+  }
+
+  Vector perform_op(const Vector& x1, const Vector& x0, const double beta) const {
+    Vector x2 = A_ * x1 - beta * x0;
+    x2.normalize();
+    return x2;
+  }
+
+  Vector perform_op(const Vector& x1) const {
+    Vector x2 = A_ * x1;
+    x2.normalize();
+    return x2;
+  }
+
+  Vector perform_op(int i) const {
+    if (accelerated_) {
+      Vector x1 = ritz_vectors_.col(i-1);
+      Vector x2 = ritz_vectors_.col(i-2);
+      return perform_op(x1, x2);
+    } else
+      return perform_op(ritz_vectors_.col(i-1));
+  }
+
+  long next_long_rand(long seed) {
+    const unsigned int m_a = 16807;
+    const unsigned long m_max = 2147483647L;
+    unsigned long lo, hi;
+
+    lo = m_a * (long)(seed & 0xFFFF);
+    hi = m_a * (long)((unsigned long)seed >> 16);
+    lo += (hi & 0x7FFF) << 16;
+    if (lo > m_max) {
+      lo &= m_max;
+      ++lo;
+    }
+    lo += hi >> 15;
+    if (lo > m_max) {
+      lo &= m_max;
+      ++lo;
+    }
+    return (long)lo;
+  }
+
+  Vector random_vec(const int len) {
+    Vector res(len);
+    const unsigned long m_max = 2147483647L;
+    for (int i = 0; i < len; i++) {
+      long m_rand = next_long_rand(m_rand);
+      res[i] = double(m_rand) / double(m_max) - double(0.5);
+    }
+    res.normalize();
+    return res;
+  }
+
+  /// Tuning the momentum beta using the Best Heavy Ball algorithm in Ref(3)
+  void setBeta() {
+    if (m_n_ < 10) return;
+    double maxBeta = beta_;
+    size_t maxIndex;
+    std::vector<double> betas = {2/3*maxBeta, 0.99*maxBeta, maxBeta, 1.01*maxBeta, 1.5*maxBeta};
+
+    Matrix tmp_ritz_vectors;
+    tmp_ritz_vectors.resize(m_n_, 10);
+    tmp_ritz_vectors.setZero();
+    for (size_t i = 0; i < 10; i++) {
+      for (size_t k = 0; k < betas.size(); ++k) {
+        for (size_t j = 1; j < 10; j++) {
+          // double rayleighQuotient;
+          if (j <2 ) {
+            Vector x0 = random_vec(m_n_);
+            Vector x00 = random_vec(m_n_);
+            tmp_ritz_vectors.col(0) = perform_op(x0, x00, betas[k]);
+            tmp_ritz_vectors.col(1) =
+                perform_op(tmp_ritz_vectors.col(0), x0, betas[k]);
+          }
+          else {
+            tmp_ritz_vectors.col(j) =
+                perform_op(tmp_ritz_vectors.col(j - 1),
+                           tmp_ritz_vectors.col(j - 2), betas[k]);
+          }
+          const Vector x = tmp_ritz_vectors.col(j);
+          const double up = x.transpose() * A_ * x;
+          const double down = x.transpose().dot(x);
+          const double mu = up / down;
+          if (mu * mu / 4 > maxBeta) {
+            maxIndex = k;
+            maxBeta = mu * mu / 4;
+            break;
+          }
+        }
+      }
+    }
+
+    beta_ = betas[maxIndex];
+  }
+
+  void perturb(int i) {
```

Comment:
You can replace this entire function with a single call to Eigen Vector's Random() static construtor